### ISSUE TITLE
ansible_connection: network_cli

### ISSUE THREAD

**Issue by nunoenpereira** [ansible/ansible#45681]
##### SUMMARY

I cannot find a way to make the variable "ansible_connection: network_cli" work in 2.6.4 and 2.7.0rc2. It works on 2.5.1 though. I verified this by calling ansible through python (upgraded and downgraded through pip3 to test).

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

Ansible network variables.

##### ANSIBLE VERSION

```paste below
ansible 2.5.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/<removed>/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/<removed>/.local/lib/python3.6/site-packages/ansible
  executable location = /home/<removed>/.local/bin/ansible
  python version = 3.6.5 (default, Apr  1 2018, 05:46:30) [GCC 7.3.0]
```

##### CONFIGURATION

```paste below
<empty output>
```

##### OS / ENVIRONMENT

ASAv on Azure cloud.

##### STEPS TO REPRODUCE

Copy the content below to files with the corresponding names. Place them all in the same folder, with the exception of the group vars file, which must be in the group_vars folder. Call main.py on the main folder.

```yaml
inventory.yml:
all:
    children:
        firewalls:
            hosts:
                <removed>:

firewalls.yml in group_vars folder:
---
ansible_connection: network_cli
ansible_network_os: asa
ansible_user: "{{username}}"
ansible_ssh_pass: "{{password}}"
ansible_become: yes
ansible_become_method: enable
ansible_become_pass: "{{enpassword}}"
ansible_ssh_common_args: "-o KexAlgorithms=+diffie-hellman-group1-sha1"

playbook.yml:
---
- hosts: firewalls
  vars_prompt:
  - name: username
    prompt: "Username"
    private: no
  - name: password
    prompt: "Password"
    private: yes
  - name: enpassword
    prompt: "Enable password"
    private: yes
  tasks:
  - name: show hostname
    asa_command:
      commands:
        - show hostname
    register: sh_result
  - debug:
      var: sh_result.stdout_lines

main.py:
from ansible.cli.playbook import PlaybookCLI
inventory = 'inventory.yml'
playbook = 'playbook.yml'
cli = PlaybookCLI(args=['ansible-playbook', '-i', inventory, playbook])
cli.parse()
cli.run()
```

##### EXPECTED RESULTS

Fail on versions 2.6.4 and 2.7.0 with the message: "invalid literal for int() with base 10: '(dp1'".
Success on version 2.5.1.

##### ACTUAL RESULTS

I don't know how to do this in python. Tell me if it is really needed and I will try to do it.

```paste below
invalid literal for int() with base 10: '(dp1'
```

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Qalthos** [comment]
Would it be possible for you to instead just run `ansible-playbook -i inventory.yml playbook.yml -vvv` and report the result of that back? if not then appending `'-vvv'` to `args` and providing the version output from the _current_ version of ansible you are trying to use would be almost as good.

My immediate guess  based on the information you have provided is that you have two versions of Ansible installed. One globally (perhaps from a package manager), and another installed via `pip3 install --user`. So now we have two versions of `ansible-connection` on the system, and the one installed to the system likely has precedence in PATH. This is further exacerbated by running the playbook as a python script, as it can't do the normal cheat of looking in `os.path.dirname(sys.argv[0])`. This is an issue that has been around for a while, and should have a satisfactory resolution in 2.8, however in the meantime, there are a few possible solutions right now.

* The simplest solution is probably to run ansible with `PATH="<absolute path of desired ansible>/bin:$PATH"` which will ensure that the correct ansible is picked first. That should almost certainly be `/home/<removed>/.local/bin`, as the version installed with `--user` will be masking anything on your system

* On that note, it is probable that the version installed with `--user` is _not_ the one you want. I have only been able to reproduce this when a version of ansible < 2.6.3 was installed with `--user`, and a version higher than that was installed to the system. Given the output of `ansible --version` in your issue, this seems likely to be the case. In that instance `pip3 uninstall ansible` should remove that installation (check the output to make sure it's removing the one in `/home/<removed>/.local/`!) and then starting a new terminal window should leave you with only the system installed version, and there shouldn't be any conflicts anymore

* Alternately, a virtual environment is exactly the sort of tool to deal with this sort of situation, where you want things installed that have different versions from what the system provides. This will give you the option of `pip install`ing whichever version you want without worrying about PATH munging or what version is installed on the system.

That's as much advice as I can give without knowing more about what is installed on your system and why. At some point, you have two different versions of ansible installed, and the call to ansible-connection is picking a newer one than the ansible that is running your playbook. Beyond that, most of this is speculation or supposition, so if I've made some incorrect assumptions, let me know.

needs_info

**Comment by nunoenpereira** [comment]
@Qalthos, regarding "you have two versions of Ansible installed", it was on purpose because, being new to ansible, I wanted to check how different it is to use ansible through the cli versus using it through python. However, when I started to have the reported issue, I noticed that the default installations deploy different versions. Then, I tried to level the "apt one" with the "pip3 one", first at ansible 2.6.4 and then at 2.5.1. "ansible_connection: network_cli" was only working in 2.5.1 by then.

After reading your post, I tried to remove both with apt and pip3 to make a clean install, but I noticed remaining "residual configurations", when doing "apt list". I think something was being left over during the ansible upgrades and downgrades to see if it was a version problem.

I am still not sure what the problem was, but as I am using WSL Ubuntu on Windows 10, I just reinstalled WSL, ansible 2.6.4 with apt and ansible 2.6.4 with pip3, and it is working ok now.

Thank you!
